### ISSUE TITLE
Intermittent Segfault In CI For Python 3.7/PySide2 Bindings

### ISSUE THREAD

**Issue by j9ac9k** [pyqtgraph/pyqtgraph#1825]
While working on #1818 I noticed being able to routinely trigger a segfault on Python 3.7/PySide2.  Initially I could reliably trigger it on macOS, but occasionally on Linux, and rarely on Windows as well.

I broke up the pytest call to test examples and tests directory separately, and didn't see a segfault for a while until I was just about to merge.

Seeing that same issue on another PR, I decided to take a look in the CI history to see when this issue first came, and it would appear it was introduced as part of #1796 here is the first CI run that saw this failure: https://github.com/pyqtgraph/pyqtgraph/actions/runs/877375417

Unfortunately due to the number of rebases, it's a bit tough to track down; but I'm fairly sure this is the first commit:

https://github.com/pyqtgraph/pyqtgraph/commit/5955a6f3d2955a1caed59b48a72739548d7d799d

Of the changes that were made here, that made their way to `master` were 

* the chances of the `struct` construction (so we could get rid of those pesky `[1:-1]` indexing
* introduction of the `skipFiniteCheck` optimization
* the use of QPolygonF construction in the use-case of `connect='all'`.
* the change in the QByteArray construction
*  adding `test_arrayToQPath` to the test suite

Given the substantial number of changes made during this commit that are still persistent in master, I'm willing to bet that the segfault started here.

@pijyoi is there anything here that catches your eye, the thing that catches my eye is the change of...

```python
    try:
        buf = QtCore.QByteArray.fromRawData(path.strn)
    except TypeError:
        buf = QtCore.QByteArray(bytes(path.strn))
    except AttributeError:
        # PyQt6 raises AttributeError
        buf = QtCore.QByteArray(path.strn, path.strn.nbytes)
```

to 

```python
    path.strn = backstore
    if QT_LIB == "PyQt6":
        # due to issue detailed here:
        # https://www.riverbankcomputing.com/pipermail/pyqt/2021-May/043942.html
        buf = QtCore.QByteArray(path.strn, len(path.strn))
    else:
        buf = QtCore.QByteArray(path.strn)
```

Would sure hate for the QPolygonF construction to be triggering a segfault here :|

EDIT: forgot to say why I am suspiscious of that case.  This issue reminds me of this:

https://github.com/pyqtgraph/pyqtgraph/pull/1420#issuecomment-716988730

segfault was avoided by this diff: https://github.com/pyqtgraph/pyqtgraph/pull/1420/commits/45a3ea240cdb399277d4fe5d3775bde8a342f867

**Comment by pijyoi** [comment]
If you are referring to test_ScatterPlotItem.py, it has been intermittently segfaulting quite far back. https://github.com/pyqtgraph/pyqtgraph/runs/2562910342?check_suite_focus=true

Now that doesn't mean that we couldn't have added something to make it segfault even more regularly.

The current implementation for PySide2 is as follows (after stripping away the PyQt6 6.0.x branch):
```python
    path.strn = backstore
    buf = QtCore.QByteArray(path.strn)
    ds = QtCore.QDataStream(buf)
    ds >> path
    return path
```
The relevant difference is that previously, a ```bytes``` was passed to QByteArray, whereas now it's a ```bytearray```. However, what I had understood was that the conversion to ```bytes``` was only because previously the buffer was an ndarray view which was not accepted as an input argument. Furthermore, the QByteArray() constructor makes a copy of its input. Of course, here I am assuming that (PySide2 5.12, Python 3.7) behaves the same as (PySide2 5.15, Python 3.8) and (PySide6, Python 3.9).

But if there's a doubt here, you could change ```path.strn = backstore``` to ```path.strn = bytes(backstore)``` and see if it makes a difference.

**Comment by pijyoi** [comment]
>segfault was avoided by this diff: 45a3ea2

So it's a bug present in PySide2 5.12.6 but not in PySide2 5.15.2
```python
In [2]: QtCore.__version__
Out[2]: '5.12.6'

In [3]: bytes(QtCore.QByteArray(b'abc'))
Segmentation fault
```

```python
In [2]: QtCore.__version__
Out[2]: '5.15.2'

In [3]: bytes(QtCore.QByteArray(b'abc'))
Out[3]: b'abc'
```

What we currently do (constructing from a bytearray) seems fine though.
```python
In [4]: QtCore.__version__
Out[4]: '5.12.6'

In [5]: QtCore.QByteArray(bytearray(b'abc\0def'))
Out[5]: PySide2.QtCore.QByteArray(b'abc\x00def')
```

**Comment by j9ac9k** [comment]
Hi @pijyoi 

 I am indeed talking about the scatter plot item segfault, but looking at the CI runs historically, on the same combination of bindings we have seen intermittent segfaults with `test_numba_makeARGB_against_generated_references` as well.  This makes me think the issue isn't so much in those tests directly, but perhaps something happening during garbage collection as the test suite is rolling along.  I don't understand the process anywhere near well enough to speculate as to the reasons why this is intermittent; but I suspect those same reasons is why I have not been able to reproduce locally.

Thanks for linking the other CI run that failed.  I hadn't noticed this test combination fail before; if we kept testing PySide6 6.0.4 wonder if we would see that more often, that binding was relatively short lived given that PySide 6.1.0 came out a week later.  I installed that configuration in an attempt to reproduce but was unfortunately unable to do so.

Another thing I've attempted to do to reproduce locally is run the test suite on a far more resource limited linux VM (2 GB of RAM, 2 CPU threads) but had no luck reproducing the issue there either.
